### PR TITLE
Feedback

### PR DESCRIPTION
**Pull request by nihirv**
:wave:! GitHub Classroom created this pull request as a place for your teacher to leave feedback on your work. It will update automatically. **Don’t close or merge this pull request**, unless you’re instructed to do so by your teacher.

In this pull request, your teacher can leave comments and feedback on your code. Click the **Subscribe** button to be notified if that happens.

Click the **Files changed** or **Commits** tab to see all of the changes pushed to `master` since the assignment started. Your teacher can see this too.

<details>
<summary>
<strong>Notes for teachers</strong>
</summary>

Use this PR to leave feedback. Here are some tips:

- Click the **Files changed** tab to see all of the changes pushed to `master` since the assignment started. To leave comments on specific lines of code, put your cursor over a line of code and click the blue **+** (plus sign). To learn more about comments, read “[Commenting on a pull request](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request)”.
- Click the **Commits** tab to see the commits pushed to `master`. Click a commit to see specific changes.
- If you turned on autograding, then click the **Checks** tab to see the results.
- This page is an overview. It shows commits, line comments, and general comments. You can leave a general comment below.

For more information about this pull request, read “[Leaving assignment feedback in GitHub](https://classroom.github.com/help/leaving-feedback-in-github)”.
</details>

Subscribed: @Farhad1353

### CODE REVIEW

File: AB_1st_stage.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
+import sys                    # import libraries
```
Comment (life-efficient) [inline AB_1st_stage.py]: The name of this file is not at all clear what it is. AB_1st_stage.py

File: AB_1st_stage.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+my_train_File, my_test_File = read_csv_files(path_train_file, path_test_file) # reading the csv files using pandas
+                                                                       
+header, X_train, X_validation, Y_train, Y_validation\
+ = split_train_validation(my_train_File, split_rate=0.25) # splitting the train.csv to train and validation arrays
+
+X_test, Y_test = split_features_labels(my_test_File) # splitting the test.csv file to features and label arrays
+                                                                                    
+###################################################################################################################
+################### This part is using AdaBoost in Regression form(Part 2 of the Project) #########################
```
Comment (life-efficient) [inline AB_1st_stage.py]: Please stick with a consistent comment scheme. The first comment says what part of the project in the start of the sentence, this one says it in () at the end. Why is that? Always be consistent.

File: AB_1st_stage.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+###################################################################################################################
+best_regr_score_adaboost = 0 # variable which shows the best scoring AdaBoost model
+progress_adaboost = 0  # variable which indicates how far have we processed our AdaBoost models
```
Comment (life-efficient) [inline AB_1st_stage.py]: Great commenting. Although it is obvious that this is a variable. 

Replace 
`# variable which allows us to X`
with 
`# X`

File: AB_1st_stage.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+regr_score_adaboost = 0 # variable which shows the score for each of the AdaBoost model
+
+n_estimator_strarting_point = 10   # setting the range for n_estimators for different AdaBoost models
```
Comment (life-efficient) [inline AB_1st_stage.py]: variable misspelling

File: AB_1st_stage.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+n_estimator_ending_point = 100
+n_estimator_step_size = 10 
+
+learning_rate_starting_point = 7   # setting the range for learning_rates for different AdaBoost models
+learning_rate_ending_point = 100
+learning_rate_step_size = 3            
+
+number_of_n_estimators \
```
Comment (life-efficient) [inline AB_1st_stage.py]: number_of_n_estimators? shouldn't this be n_estimators or number_of_estimators

File: Data/AdaBoost_reg.csv [hunk 6]
```diff
@@ -0,0 +1,2 @@
+30
```
Comment (life-efficient) [inline Data/AdaBoost_reg.csv:1]: again, no idea what is in this file because the name doesn't tell me anything. what is reg???

File: GB_1st_stage.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+learning_rate_starting_point = 3   # setting the range for learning_rate for different GradientBoosting models
+learning_rate_ending_point = 50
+learning_rate_step_size = 5     
+
+max_depth_starting_point = 3    # setting the range for max_depth for different GradientBoosting models
+max_depth_ending_point = 10
+max_depth_step_size = 2           
+
+number_of_n_estimators \
```
Comment (life-efficient) [inline GB_1st_stage.py]: not clear what this does. please comment. im sure it does not need to be this complicated, but im not sure what you're trying to do

File: Library/utils.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
+import sklearn.datasets
```
Comment (life-efficient) [inline Library/utils.py:1]: Can you please add docstrings to these functions to explain what they do. Do you know?

Comment (life-efficient) [reply Library/utils.py:1]: Also, please comment the file

File: README.md [hunk 9]
```diff
@@ -0,0 +1,8 @@
+# Ensemble-optimisation-using-hyperparameter-and-architecture-search
+This is an ensemble project to establish different Red Wines quality based on 11 different features for each wine given.
+
+In order to execute this project you can run __Main.py__ for a fast result(using Ctrl+F5).
```
Comment (life-efficient) [inline README.md]: Remove Ctrl + F5. Depends on what IDE you use.

File: README.md [hunk 10]
```diff
@@ -0,0 +1,8 @@
...
+However, for better visiualization, comments and flowchart diagram I'll certainly recommend to run __Main__.ipynb.
+
+There is 3 more .py codes that can be executed as a first stage of the ensemble model(You do not have to run this, as it has been executed at least once each and did send the necessary info to CSV files(in Data subfolder) for 2nd stage of the ensemble process) : RF_1st_stage.py, AB_1st_stage.py and GB_1st_stage. Running this 3 codes will overwrite the previous data in RandomForest_reg.csv, AdaBoost_reg.csv and GradientBoosting_reg.csv.
```
Comment (life-efficient) [inline README.md]: Call them programs, not codes. 

What is the first stage? Nobody knows.

Comment (life-efficient) [reply README.md]: Explain what your csv files are. Explain the structure of the project in general so it's easy for someone else to understand what's going on.

Comment (life-efficient) [reply README.md]: Put all of your visualisations in here so it's easy for someone to see a summary of your work.

**Review by life-efficient — CHANGES_REQUESTED** [review]

File: __Main__.ipynb [hunk 11]
```diff
@@ -0,0 +1,904 @@
+{
```
Comment (life-efficient) [inline __Main__.ipynb]: <p>realted -&gt; related. Not sure related  is the best word... it's more than related to those features... it IS those features.</p><p> </p><p>Might be better to call df.describe after loading the data rather than naming all of the features manually here. </p><p> </p><p>More details of the dataset: number of examples, etc</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core-Assignments/project-1-Farhad1353/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (life-efficient) [inline __Main__.ipynb]: <p>what is this doing here? Isnt it also in your utils?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core-Assignments/project-1-Farhad1353/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (life-efficient) [inline __Main__.ipynb]: <p>Not a useful plot. Look closely though, why are there repeating patterns in the x-direction?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core-Assignments/project-1-Farhad1353/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (life-efficient) [inline __Main__.ipynb]: <p>Don't print these params manually, print them in code in case they change -_-</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core-Assignments/project-1-Farhad1353/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (life-efficient) [inline __Main__.ipynb]: <p>stop saying "variable for..." in every line's comment. It's obvious that it is a variable lol. The comment should explain WHY, not what</p><p> </p><p>Why have you plotted this? Answer to yourself before reading on. </p><p> </p><p>This plot is useless. No insight is given by knowing the number of the model. This graph makes it look like accuracy is a function of model number, which it is not, because i could equally have shuffled the order of which hyperparams i tried (or even more simply, done the loops in the opposite order).</p><p> </p><p>I think what you mean to plot is a surface plot where height is the accuracy/loss and max_depth varies along the x-axis, and max_features varies along the y-axis.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core-Assignments/project-1-Farhad1353/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (life-efficient) [inline __Main__.ipynb]: <p>Again, useless plot. Replace it with a surface plot.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core-Assignments/project-1-Farhad1353/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (life-efficient) [inline __Main__.ipynb]: <p>Please never make another one of these plots lol</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core-Assignments/project-1-Farhad1353/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (life-efficient) [inline __Main__.ipynb]: <p>What is first stage??? Never saw that defined anywhere</p><p> </p><p>Again, dont print these vars manually. Print them in a code cell.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core-Assignments/project-1-Farhad1353/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (life-efficient) [inline __Main__.ipynb]: <p>work on your variable names.</p><p> </p><p>When i see a variable called gradientboost_models i think it should be a list of models. Instead it contains accuracies. Guess what I would suggest as a better variable name? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core-Assignments/project-1-Farhad1353/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (life-efficient) [inline __Main__.ipynb]: <p>Great! Stacking is a great approach. You should also have tried bagging. </p><p> </p><p>I like how comprehensive you have been. We are trying LOTS of permutations of models now, so this is a good search through hyperparameter space!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core-Assignments/project-1-Farhad1353/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

Comment (life-efficient) [inline __Main__.ipynb]: <p>Nice work using the validation set! Lot's of people forgot this lol. Remember why it is necessary?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core-Assignments/project-1-Farhad1353/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (life-efficient) [inline __Main__.ipynb]: <p>Please dont ask me to scroll to another part of the notebook to find the things which you are meant to be showing in this section.</p><p>Can you combine all of your results into a single graph (do not plot accuracy against model number).</p><p> </p><p>Why didnt you remove these outliers and see if that was the problem? If you didn't, then you can't claim that was the problem with certainty.</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core-Assignments/project-1-Farhad1353/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

Comment (life-efficient) [inline __Main__.ipynb]: <p>Good stuff. Some of the other students used other methods to extract feature importance too. You should add those in!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core-Assignments/project-1-Farhad1353/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Comment (life-efficient) [inline __Main__.ipynb]: <p>Wow i love this. Great visualisation! Put this at the top of the readme.md</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AI-Core-Assignments/project-1-Farhad1353/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

File: AB_1st_stage.py (hunk 1 above)
Comment (Farhad1353) [reply AB_1st_stage.py]: Hi, 
Sorry I should have named it AdaBoost_1st_stage.py instead

File: __Main__.ipynb (hunk 11 above)
Comment (Farhad1353) [reply __Main__.ipynb]: <p>Hi Harry, </p><p>This is not part of the project.</p><p>I did add this after the submitting date, as my plan was gradually upgrade the program.</p><p>I am about to delete this</p>

Comment (Farhad1353) [reply __Main__.ipynb]: <p>My expectation was to see a repeating patterns.</p><p>At 1st stage of the Random Forest,I did implement nested loop(with three loops which changes the hyper-parameters. </p><p>So as at some point when we have the same figure for the two inner loops but different value for the outer loop I expect to see some sort of repeating pattern.</p><p>I understand, I need to learn a smarter approach to find a better accuracy score. </p><p>So, I'm not justifying my method is good. I am only explaining why based on the method I've chosen we should expect a repeating pattern.</p>

Comment (Farhad1353) [reply __Main__.ipynb]: <p>I understand you are 100% correct about this.</p><p>However, the only reason, I did print this is, I did not want you you to spend approximately 5 minutes each for the 1st stage of Random Forest , Ada Boost and Gradient Boost prior to executing the 2nd stage(final stage) of the program. </p>

Comment (Farhad1353) [reply __Main__.ipynb]: <p>I think I should have presented many small plots instead of this one plot.</p><p>Also, I should have assign sensible axis-description for each sub-plot.</p>

Comment (Farhad1353) [reply __Main__.ipynb]: <p>Can I use several tiny sub-plots with proper description instead of surface plot?</p><p> </p><p>Correction : I just thought about this, I think I can get a better surface plot visualization(probably 4 or 5 plots) instead of too many tiny subplots with(2 variables only).</p>

Comment (Farhad1353) [reply __Main__.ipynb]: <p>OK</p>

Comment (Farhad1353) [reply __Main__.ipynb]: <p>I need to add a better description in README.</p><p>OK, I don't. </p>

Comment (Farhad1353) [reply __Main__.ipynb]: <p>I should have used index and score in the name of the variable instead.</p><p> However, I intentionally wanted to create a list for the scores and indices </p>

Comment (Farhad1353) [reply __Main__.ipynb]: <p>Great thanks.</p><p>I'm a bit disappointed that at the time I did submit this project, I still had difficulty installing better packages like Xgboost, Catboost and  lightgboost.</p><p>Finally, I have managed to install them and successfully implemented them.</p><p>I am very impressed with the CatBoost.</p>

Comment (Farhad1353) [reply __Main__.ipynb]: <p>Certainly,</p><p>I like the validation set, as it helps me to see how the model is(maybe overfit or underfit) at the same time that I'm training the model and helps to stop or change the direction of training if needed.</p>

Comment (Farhad1353) [reply __Main__.ipynb]: <p>I'll be moving this to the same section.</p><p> </p><p>I'm not certain the reason of the problem was the outliers, I'm probably incorrect and will look at it closely asap.</p>

Comment (Farhad1353) [reply __Main__.ipynb]: <p>I certainly need to learn more methods for this purpose.</p><p>Also, I have only used my knowledge based on Chapter 1(ML). Perhaps, if I did use some of the methods from Chapter 2(DS), I could implement better methods and visualization to the project.</p>

Comment (Farhad1353) [reply __Main__.ipynb]: <p>I just did put the link of this image in the readme.md file as I couldn't find a way to put the actual image of the flowchart in the readme.</p>

**Comment by Farhad1353** [comment]
<p>I think I could presented a better result if I did use my learning from chapter 2.</p><p>However, I feel I am still long way from where I like to be.</p><p>I hope during this week(where we are not going to have any new topic) I can look at my weaknesses carefully and prepare myself for the remaining of the course.</p><p>Also, the mini interviews with Nihir, Harry and Joao on Saturday(3/10/20) certainly helped me to understand I'm not as good as I thought(the way I thought I know everything we studied so far).</p><p> </p><p>I love to see more of this mini interviews as it will show partially where is my weakness. </p><p>I need a bit pressurizing to put me in a right path</p>
